fendermint/vm/interpreter/src/chain.rs

```diff
@@ -214,39 +223,67 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDownExec(_) => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownExec(p) => {
+                    let validator_set =
+                        atomically_or_err::<_, fendermint_vm_topdown::Error, _>(|| {
+                            provider.validator_set(p.height as u64)
+                        })
+                        .await?
+                        .ok_or_else(|| {
+                            anyhow!("cannot find validator set for block: {}", p.height)
+                        })?;
+
+                    let finality = IPCParentFinality {
+                        height: p.height as u64,
+                        block_hash: p.block_hash,
+                    };
+                    let msg = parent_finality_to_fvm(finality.clone(), validator_set)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+
+                    atomically_or_err::<_, fendermint_vm_topdown::Error, _>(|| {
+                        provider.set_new_finality(finality.clone())
+                    })
+                    .await?;
```

Comment:
It would be really confidence inspiring to root out all the cases where an operation on the provider _cannot fail_ and remove the `Result`. Why can this fail so bad that it can crash Fendermint?